### PR TITLE
[#1078] Detect periods contained in lib names inside prefix lists

### PR DESCRIPTION
**Pull request by jysandy**
Fixes https://github.com/clj-kondo/clj-kondo/issues/1078.

Output of `script/diff`:
```
Linting and writing output to /tmp/clj-kondo-diff/branch.txt
Cloning into 'clj-kondo'...
remote: Enumerating objects: 82, done.
remote: Counting objects: 100% (82/82), done.
remote: Compressing objects: 100% (65/65), done.
remote: Total 9754 (delta 29), reused 48 (delta 11), pack-reused 9672
Receiving objects: 100% (9754/9754), 11.10 MiB | 5.71 MiB/s, done.
Resolving deltas: 100% (5623/5623), done.
Linting and writing output to /tmp/clj-kondo-diff/master.txt
1c1
< "Elapsed time: 10.222993 msecs"
---
> "Elapsed time: 9.500667 msecs"
1966c1966
< clojure/tools/reader.clj:18:53: warning: use alias or :refer [char desugar-meta ex-info? namespace-keys numeric? reader-conditional second' tagged-literal whitespace?]
---
> clojure/tools/reader.clj:18:53: warning: use alias or :refer [char desugar-meta ex-info? namespace-keys numeric? second' whitespace?]
2253c2253
< inlined/clj_kondo/impl/toolsreader/v1v2v2/clojure/tools/reader.clj:16:87: warning: use alias or :refer [char desugar-meta ex-info? namespace-keys numeric? reader-conditional second' tagged-literal whitespace?]
---
> inlined/clj_kondo/impl/toolsreader/v1v2v2/clojure/tools/reader.clj:16:87: warning: use alias or :refer [char desugar-meta ex-info? namespace-keys numeric? second' whitespace?]
2693c2693
< linting took 11818ms, errors: 400, warnings: 2291
---
> linting took 12279ms, errors: 400, warnings: 2291
```

### CODE REVIEW

File: test/clj_kondo/main_test.clj [hunk 1]
```diff
@@ -2529,6 +2529,36 @@ foo/foo ;; this does use the private var
 
 " {:linters {:unsorted-required-namespaces {:level :warning}}})))))
 
+(deftest require-prefix-form-containing-periods-test
```
Comment (jysandy) [inline test/clj_kondo/main_test.clj:2532]: Should I add these tests into `corpus/prefixed_libspec.clj` instead?

Comment (jysandy) [reply test/clj_kondo/main_test.clj:2532]: Then they'll be tested under https://github.com/clj-kondo/clj-kondo/blob/master/test/clj_kondo/main_test.clj#L487

Comment (jysandy) [reply test/clj_kondo/main_test.clj:2532]: Not entirely sure that's necessary however, it might be simpler to keep them separate.

Comment (borkdude) [reply test/clj_kondo/main_test.clj:2532]: Splitting it out either way is fine with me.

File: test/clj_kondo/main_test.clj [hunk 2]
```diff
@@ -2529,6 +2529,36 @@ foo/foo ;; this does use the private var
...
+  (is (= '({:col     31
+            :file    "<stdin>"
+            :level   :error
+            :message "Lib names inside prefix lists must not contain periods."
+            :row     3})
+         (lint! "(ns baz
+                   (:require [clj-kondo.impl.analyzer
+                              foo.bar]))"
+                {:linters {:unused-namespace {:level :off}}}))))
+
```
Comment (borkdude) [inline test/clj_kondo/main_test.clj:2561]: Can you explain what the difference in test is between the first and second one? Maybe clarify using `testing`.

Comment (jysandy) [reply test/clj_kondo/main_test.clj:2561]: The first one contains one valid lib name (`babashka`), whereas the second one does not. On second thought though, it doesn't seem like a valuable test case. I think I'll get rid of it, and replace it with a test case that contains multiple invalid lib names.

**Comment by jysandy** [comment]
I also ran into a case where nested prefix forms are not accepted by the compiler, but clj-kondo doesn't detect them. 
Ex: 
```
(ns quux
  (:require [clj-kondo.impl 
             [analyzer
              [foo]
              [bar]]]))
```
I'll include a fix for this.

File: test/clj_kondo/main_test.clj (hunk 2 above)
Comment (jysandy) [reply test/clj_kondo/main_test.clj]: This is changed.

File: test/clj_kondo/main_test.clj [hunk 3]
```diff
@@ -486,7 +486,12 @@ foo/foo ;; this does use the private var
 
 (deftest prefix-libspec-test []
   (assert-submaps
-   '({:file "corpus/prefixed_libspec.clj",
+   '({:col 14
+      :file "corpus/prefixed_libspec.clj"
+      :level :error
+      :message "nested prefix lists are not allowed."
```
Comment (jysandy) [inline test/clj_kondo/main_test.clj:492]: Not sure if this is a helpful error message. Trying to think of a better one.

File: src/clj_kondo/impl/analyzer/namespace.clj [hunk 4]
```diff
@@ -41,19 +41,34 @@
...
-                                       (symbol (str (when prefix (str prefix "."))
-                                                    (first form)))
-                                       f))
-                  (rest children))
+          (do (when prefix
+                (findings/reg-finding! ctx (node->line (:filename ctx)
+                                                       libspec-expr
+                                                       :error
+                                                       :syntax
+                                                       "nested prefix lists are not allowed.")))
```
Comment (jysandy) [inline src/clj_kondo/impl/analyzer/namespace.clj:49]: Not sure if this is a helpful error message. Trying to think of a better one.

Comment (borkdude) [reply src/clj_kondo/impl/analyzer/namespace.clj:49]: If you are using punctuation, can you also start the sentence with an uppercase? What is the case you are finding here?

Comment (borkdude) [reply src/clj_kondo/impl/analyzer/namespace.clj:49]: In general I try to use the same message as the Clojure compiler itself does.

Comment (jysandy) [reply src/clj_kondo/impl/analyzer/namespace.clj:49]: This is the case of having nested prefix forms, which doesn't seem to be allowed. Ex:
```
(ns quux
    (:require [clj-kondo.impl
                       [analyzer
                         [foo]
                         [bar]]]))
```
This results in a spec failure, the compiler doesn't seem to have a custom error message for it.
```
Syntax error macroexpanding clojure.core/ns at (REPL:1:1).
((:require [foo [bar [baz :as b]] [baz :as c]])) - failed: Extra input spec: :clojure.core.specs.alpha/ns-form
```

Comment (borkdude) [reply src/clj_kondo/impl/analyzer/namespace.clj:49]: So the issue is that nested libspecs actually only support one nesting?

Comment (jysandy) [reply src/clj_kondo/impl/analyzer/namespace.clj:49]: @borkdude yes, that's right.

Comment (borkdude) [reply src/clj_kondo/impl/analyzer/namespace.clj:49]: > Prefix Lists

> It's common for Clojure code to depend on several libs whose names have
the same prefix. When specifying libs, prefix lists can be used to reduce
repetition. A prefix list contains the shared prefix followed by libspecs
with the shared prefix removed from the lib names. After removing the
prefix, the names that remain must not contain any periods.

A nicer message may be "Prefix lists can only have two levels." perhaps?

Comment (jysandy) [reply src/clj_kondo/impl/analyzer/namespace.clj]: Sounds good to me, I've changed it.

**Pull request merged**